Firestore/Source/API/FIRQuery.m

```diff
@@ -256,6 +256,82 @@ - (FIRQuery *)queryWhereFieldPath:(FIRFieldPath *)path isGreaterThanOrEqualTo:(i
                                  value:value];
 }
 
+- (FIRQuery *)queryWherePredicate:(NSPredicate *)predicate {
+  if ([predicate isKindOfClass:[NSComparisonPredicate class]]) {
+    // Comparison predicate
+    NSComparisonPredicate *comparison = (NSComparisonPredicate *)predicate;
+    if (comparison.comparisonPredicateModifier != NSDirectPredicateModifier) {
+      FSTThrowInvalidArgument(@"Invalid query. Predicate cannot have an "
+                              "aggregate modifier.");
+    }
+    NSString *path;
+    id value = nil;
+    if (comparison.leftExpression.keyPath == nil &&
+        comparison.rightExpression.keyPath == nil) {
+      FSTThrowInvalidArgument(@"Invalid query. Predicate comparison must have "
+                              "a key path on either side of its expression.");
+    } else if (comparison.leftExpression.keyPath == nil) {
+      path = comparison.rightExpression.keyPath;
+      value = comparison.leftExpression.constantValue;
+    } else {
+      path = comparison.leftExpression.keyPath;
+      value = comparison.rightExpression.constantValue;
+    }
+    if (value == nil) {
+      FSTThrowInvalidArgument(@"Invalid query. Predicate comparison must have "
+                              "a constant on its expression.");
+    }
+    switch (comparison.predicateOperatorType) {
+      case NSEqualToPredicateOperatorType:
+        return [self queryWhereField:path isEqualTo:value];
+        break;
+      case NSLessThanPredicateOperatorType:
+        return [self queryWhereField:path isLessThan:value];
+        break;
+      case NSLessThanOrEqualToPredicateOperatorType:
+        return [self queryWhereField:path isLessThanOrEqualTo:value];
+        break;
+      case NSGreaterThanPredicateOperatorType:
+        return [self queryWhereField:path isGreaterThan:value];
+        break;
+      case NSGreaterThanOrEqualToPredicateOperatorType:
+        return [self queryWhereField:path isGreaterThanOrEqualTo:value];
+        break;
+      case NSCustomSelectorPredicateOperatorType:
+        FSTThrowInvalidArgument(@"Invalid query. Custom predicate filters are "
+                                "not supported.");
+        break;
+      default:
+        FSTThrowInvalidArgument(@"Invalid query. Predicate does not support "
```

Comment:
This "Predicate does not support" phrasing isn't quite right. `NSPredicate` does support these things and this phrasing seems to suggest that the predicate itself doesn't support these.

The other phrasing just above is much better: "Custom predicate filters are not supported" is a better example.

Maybe this one could be

```
Operator type %lu is not supported.
```